### PR TITLE
Spark: Support spec ID and partition metadata columns

### PR DESCRIPTION
**Pull request by aokolnychyi**
This WIP PR adds support for `_spec` and `_partition` metadata columns so that we can project them for writing deltas.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java [hunk 1]
```diff
@@ -49,26 +54,40 @@ private MetadataColumns() {
   public static final String DELETE_FILE_ROW_DOC = "Deleted row values";
 
   private static final Map<String, NestedField> META_COLUMNS = ImmutableMap.of(
+      SPEC.name(), SPEC,
       FILE_PATH.name(), FILE_PATH,
       ROW_POSITION.name(), ROW_POSITION,
       IS_DELETED.name(), IS_DELETED);
 
-  private static final Set<Integer> META_IDS = META_COLUMNS.values().stream().map(NestedField::fieldId)
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetadataColumns.java:56]: Had to change this as the partition type is not static and is handled specially.

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java [hunk 2]
```diff
@@ -49,26 +54,40 @@ private MetadataColumns() {
...
+  );
 
   public static Set<Integer> metadataFieldIds() {
     return META_IDS;
   }
 
-  public static NestedField get(String name) {
-    return META_COLUMNS.get(name);
+  public static NestedField metadataColumn(Table table, String name) {
+    if (name.equals(PARTITION_COLUMN_NAME)) {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetadataColumns.java:75]: I kept the logic case sensitive as before but we maybe should reconsider it at some point.

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 3]
```diff
@@ -246,4 +246,45 @@ private static boolean equivalentIgnoringNames(PartitionField field, PartitionFi
...
+
+  /**
+   * Adapts the provided partition data to match the table partition type built using {@link #partitionType(Table)}.
+   *
+   * @param partitionType a table partition type that includes partition fields from all specs
+   * @param spec a partition spec
+   * @param partition a struct containing partition data that belongs to the spec
+   * @return the reconciled partition type
+   */
+  public static StructLike reconcilePartition(StructType partitionType, PartitionSpec spec, StructLike partition) {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/Partitioning.java:258]: Naming ideas?

File: core/src/main/java/org/apache/iceberg/util/PartitionUtil.java [hunk 4]
```diff
@@ -51,6 +53,20 @@ private PartitionUtil() {
...
 
+    // add _spec
+    idToConstant.put(
+        MetadataColumns.SPEC.fieldId(),
+        convertConstant.apply(Types.IntegerType.get(), task.file().specId()));
+
+    // add _partition
+    if (partitionType.fields().isEmpty()) {
+      // use null as some query engines may not be able to handle empty structs
+      idToConstant.put(MetadataColumns.PARTITION_COLUMN_ID, null);
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/util/PartitionUtil.java:64]: Using null for unpartitioned tables.

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java [hunk 5]
```diff
@@ -41,21 +38,14 @@
 
   private static final long serialVersionUID = 1L;
 
-  private final TableLoader tableLoader;
```
Comment (aokolnychyi) [inline flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java:44]: All Flink and MR changes will go in separate PRs.

File: spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java [hunk 6]
```diff
@@ -155,6 +160,24 @@ protected static Object convertConstant(Type type, Object value) {
         return ByteBuffers.toByteArray((ByteBuffer) value);
       case BINARY:
         return ByteBuffers.toByteArray((ByteBuffer) value);
+      case STRUCT:
+        StructType structType = (StructType) type;
+
+        if (structType.fields().isEmpty()) {
+          return new GenericInternalRow();
```
Comment (aokolnychyi) [inline spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:167]: We won't hit this clause with `_partition` as the passed value will be null.

File: spark3/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java [hunk 7]
```diff
@@ -149,7 +149,7 @@ private Schema schemaWithMetadataColumns() {
     // metadata columns
     List<Types.NestedField> fields = metaColumns.stream()
         .distinct()
-        .map(MetadataColumns::get)
+        .map(name -> MetadataColumns.metadataColumn(table, name))
```
Comment (aokolnychyi) [inline spark3/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java:152]: Need to pass a `Table` object to figure our the partition type.

File: spark3/src/test/java/org/apache/iceberg/spark/source/SparkTestTable.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+import org.apache.iceberg.MetadataColumns;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.spark.SparkSchemaUtil;
+import org.apache.iceberg.types.Types;
+import org.apache.spark.sql.connector.read.ScanBuilder;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+
+public class SparkTestTable extends SparkTable {
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/spark/source/SparkTestTable.java:30]: This is purely a temp solution for testing purposes until we compile against 3.2.

File: spark3/src/test/java/org/apache/iceberg/spark/source/TestSparkCatalog.java [hunk 9]
```diff
@@ -31,7 +30,14 @@
 
   @Override
   public Table loadTable(Identifier ident) throws NoSuchTableException {
-    TestTables.TestTable table = TestTables.load(Spark3Util.identifierToTableIdentifier(ident).toString());
-    return new SparkTable(table, false);
+    String[] parts = ident.name().split("\\$", 2);
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/spark/source/TestSparkCatalog.java:33]: I probably need to add a comment here. This is also for testing purposes until we consume Spark 3.2.

File: spark3/src/test/java/org/apache/iceberg/spark/source/TestSparkMetadataColumns.java [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+  @After
+  public void dropTable() {
+    TestTables.clearTables();
+  }
+
+  @Test
+  public void testSpecAndPartitionMetadataColumns() {
+    // TODO: support metadata structs in vectorized ORC reads
+    Assume.assumeFalse(fileFormat == FileFormat.ORC && vectorized);
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/spark/source/TestSparkMetadataColumns.java:115]: ORC vectorized path does not support constant structs.

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java [hunk 11]
```diff
@@ -72,12 +62,7 @@ public BaseStatistics getStatistics(BaseStatistics cachedStatistics) {
 
   @Override
   public FlinkInputSplit[] createInputSplits(int minNumSplits) throws IOException {
-    // Called in Job manager, so it is OK to load table from catalog.
-    tableLoader.open();
```
Comment (aokolnychyi) [inline flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java:76]: @stevenzwu @openinx, this is something we talked a little bit about in #2978.

After a closer look, this would actually change the behavior. As of now, the table is loaded two times: in `FlinkSource` and in `FlinkInputFormat`. After this change, we only load it in `FlinkSource` and then pass a read-only serializable copy to `FlinkInputFormat`.

What is the lifecycle of `FlinkInputFormat`? Is it alright to pass a read-only table to it that won't pick up any new snapshots or table properties?

Comment (aokolnychyi) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java:76]: As I stated above, I've put Flink and MR related changes into this PR just temporarily.

Instead of passing `Table`, I can add one more parameter and pass a partition type alongside other existing parameters. Using `Table` seems cleaner but not a blocker for my current work.

**Comment by aokolnychyi** [comment]
cc @openinx @stevenzwu @rdblue @rdsr @pvary @RussellSpitzer @flyrain @szehon-ho @karuppayya @kbendick @jackye1995

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java (hunk 11 above)
Comment (aokolnychyi) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java:76]: I had to modify a test in `TestFlinkScanSql`.

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java [hunk 12]
```diff
@@ -38,6 +38,11 @@ private MetadataColumns() {
       Integer.MAX_VALUE - 2, "_pos", Types.LongType.get(), "Ordinal position of a row in the source data file");
   public static final NestedField IS_DELETED = NestedField.required(
       Integer.MAX_VALUE - 3, "_deleted", Types.BooleanType.get(), "Whether the row has been deleted");
+  public static final NestedField SPEC = NestedField.required(
+      Integer.MAX_VALUE - 4, "_spec", Types.IntegerType.get(), "Spec ID to which a row belongs to");
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/MetadataColumns.java:42]: Should it be `_spec_id`?

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 3 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:258]: Should we use struct projections rather than copying values?

File: core/src/main/java/org/apache/iceberg/util/PartitionUtil.java [hunk 13]
```diff
@@ -51,6 +53,20 @@ private PartitionUtil() {
...
+    } else {
+      StructLike partition = Partitioning.reconcilePartition(partitionType, spec, partitionData);
+      idToConstant.put(MetadataColumns.PARTITION_COLUMN_ID, convertConstant.apply(partitionType, partition));
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/util/PartitionUtil.java:67]: Is `partitionType` derived from the read schema? It would be nice to be able to project.

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/util/PartitionUtil.java:67]: Actually, this probably doesn't matter because Spark will project just what it needs immediately and this is basically a constant for each file.

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 14]
```diff
@@ -246,4 +246,45 @@ private static boolean equivalentIgnoringNames(PartitionField field, PartitionFi
   private static boolean compatibleTransforms(Transform<?, ?> t1, Transform<?, ?> t2) {
     return t1.equals(t2) || t1.equals(Transforms.alwaysNull()) || t2.equals(Transforms.alwaysNull());
   }
+
+  /**
+   * Adapts the provided partition data to match the table partition type built using {@link #partitionType(Table)}.
+   *
+   * @param partitionType a table partition type that includes partition fields from all specs
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/Partitioning.java:253]: What is the behavior for when the partition does not contain a field that is required to create the target partition type

Like if I add a column, and then partition on that column.

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 3 above)
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:258]: coercePartition?

File: core/src/main/java/org/apache/iceberg/util/PartitionUtil.java (hunk 13 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/util/PartitionUtil.java:67]: You mean supporting schema pruning for metadata columns?

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 14 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:253]: You mean what if the table partition type (which is a union of all spec types) contains a field that is not in the struct for a particular spec?

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 3 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:258]: I'll check our current projections to see if they support going from a more specific to a more general schema.

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:258]: I'll switch to `coercePartition` too.

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 14 above)
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:253]: Yeah, We are just setting that to null right?

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java [hunk 15]
```diff
@@ -38,6 +38,11 @@ private MetadataColumns() {
...
+  public static final int PARTITION_COLUMN_ID = Integer.MAX_VALUE - 5;
+  public static final String PARTITION_COLUMN_NAME = "_partition";
+  public static final String PARTITION_COLUMN_DOC = "Partition to which a row belongs to";
```
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/MetadataColumns.java:45]: Nit: I see here that this is so that we can have a later call to `Partitioning.partitionType` (and possibly because it's optional).

But it took me a minute to see why this one is different. Possibly might consider making it more clear or adding a comment? Even just `// Partition type is not static so must be handled specially` might be good, though I know sometimes we prefer the less comments route due to comment rot etc. Up to you 🙂

File: mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java [hunk 16]
```diff
@@ -141,7 +144,7 @@
           // TODO: We do not support residual evaluation for HIVE and PIG in memory data model yet
           checkResiduals(task);
         }
-        splits.add(new IcebergSplit(conf, task, table.io(), table.encryption()));
+        splits.add(new IcebergSplit(SerializableTable.copyOf(table), conf, task));
```
Comment (kbendick) [inline mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:147]: Nit: It looks like this call to `SerializableTable.copyOf(table)` is happening in a loop. Can we pull it out?

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java [hunk 17]
```diff
@@ -49,26 +54,40 @@ private MetadataColumns() {
...
   }
 
   public static boolean isMetadataColumn(String name) {
-    return META_COLUMNS.containsKey(name);
+    return name.equals(PARTITION_COLUMN_NAME) || META_COLUMNS.containsKey(name);
   }
 
   public static boolean nonMetadataColumn(String name) {
-    return !META_COLUMNS.containsKey(name);
+    return !name.equals(PARTITION_COLUMN_NAME) && !META_COLUMNS.containsKey(name);
```
Comment (stevenzwu) [inline core/src/main/java/org/apache/iceberg/MetadataColumns.java:91]: nit: is this simpler? 
```
return !isMetadataColumn(name);
```

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java [hunk 18]
```diff
@@ -49,26 +54,40 @@ private MetadataColumns() {
   public static final String DELETE_FILE_ROW_DOC = "Deleted row values";
 
   private static final Map<String, NestedField> META_COLUMNS = ImmutableMap.of(
+      SPEC.name(), SPEC,
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/MetadataColumns.java:57]: nit: should add as the last one instead of the first one (although I know order doesn't matter in a map, better to be consistent in code)

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java [hunk 19]
```diff
@@ -49,26 +54,40 @@ private MetadataColumns() {
...
+          Partitioning.partitionType(table),
+          PARTITION_COLUMN_DOC);
+    } else {
+      return META_COLUMNS.get(name);
+    }
   }
 
   public static boolean isMetadataColumn(String name) {
-    return META_COLUMNS.containsKey(name);
+    return name.equals(PARTITION_COLUMN_NAME) || META_COLUMNS.containsKey(name);
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/MetadataColumns.java:87]: I am starting to wonder if it is better to just have a dummy column type mapping for `_partition` in `META_COLUMNS` and only separate the logic when necessary, which can avoid changes like this in quite a few places.

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 20]
```diff
@@ -246,4 +246,45 @@ private static boolean equivalentIgnoringNames(PartitionField field, PartitionFi
...
+      } else {
+        // set the value as null as the spec does not contain this partition field
+        newPartition.set(position, null);
+      }
+    }
+
+    return newPartition;
+  }
+
+  private static Map<Integer, Integer> buildFieldIdToPositionMap(PartitionSpec spec) {
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/Partitioning.java:280]: nit: since there are quite a few positions here, maybe make this name clearer as `buildFieldIdToSpecPositionMap`

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java (hunk 11 above)
Comment (jackye1995) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java:76]: not an expert in Flink, but agree this is a change of behavior that we should be cautious. Currently it seems that the input splits are created with the freshly loaded table from the serializable table loader. This means that a table can potentially be updated during the process and then generate a different array of splits comparing to using a read-only copy. 

I am not sure which behavior is correct though, it feels to me that using a read-only copy is more valid to ensure snapshot isolation. But if the Flink source is supposed to be reused to reflect always the latest view of the data, then using the table loader becomes necessary.

Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java:76]: > What is the lifecycle of FlinkInputFormat? Is it alright to pass a read-only table to it that won't pick up any new snapshots or table properties?

@aokolnychyi  I think it is ok to pass a read-only table to FlinkInputFormat. it is used in two scenarios
1) batch mode: split discovery is done once at startup
2) streaming mode: it is used only in reader. dynamic/periodical split discovery is done by `StreamingMonitorFunction`, which would require a TableLoader to refresh table state

Comment (aokolnychyi) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java:76]: Okay, I've submitted #2987 for Flink changes alone.

File: mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java (hunk 16 above)
Comment (aokolnychyi) [reply mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:147]: Good catch, @kbendick! Moved it outside of the loop in #2988.

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java (hunk 15 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MetadataColumns.java:45]: Sure, I'll add a comment above.

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java (hunk 19 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MetadataColumns.java:87]: I did that first but `META_COLUMNS` is an immutable map that does not allow null keys. I am reluctant to switch to a mutable map so I added this condition here. I hope we will be able to use `metadataColumn(table, name)` in other places so this workaround will be only part of `MetadataColumns`.

It looks ugly, though, I agree.

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java (hunk 18 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MetadataColumns.java:57]: I'll switch. Wanted to reduce the number of modified lines but probably better to follow the order of definition.

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 14 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:253]: Yeah, it will be set to null as all partition fields are optional and the table partition type is a union of all spec types.

**Comment by aokolnychyi** [comment]
I've updated this PR to focus on Spark for now. Similar functionality can be added to other query engines later.

**Comment by aokolnychyi** [comment]
I'd appreciate another review round.

**Comment by aokolnychyi** [comment]
This is no longer WIP.

File: api/src/main/java/org/apache/iceberg/util/StructProjection.java [hunk 21]
```diff
@@ -58,12 +58,32 @@ public static StructProjection create(Schema dataSchema, Schema projectedSchema)
...
   }
 
+  /**
+   * Creates a projecting wrapper for {@link StructLike} rows.
+   * <p>
+   * This projection does not work with repeated types like lists and maps.
+   *
+   * @param structType type of rows wrapped by this projection
+   * @param projectedStructType result type of the projected rows
+   * @param projectMissingFieldsAsNulls a flag whether to project missing fields as nulls
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/util/StructProjection.java:68]: You might consider renaming this to `allowMissing`. The original behavior is to fail if there is a field missing from the base type and this change really just allows that case and does the most reasonable thing -- it fills in null. But we don't necessarily want to guarantee that this will always fill in null. So I'd change the boolean to state that it allows missing fields rather than it fills in nulls.

Also, I think it is better for APIs in general to hide boolean arguments. Since this is already a factory method, why not use a name that signals missing fields are allowed? Something like `create` and `createAllowMissing`?

File: api/src/main/java/org/apache/iceberg/util/StructProjection.java [hunk 22]
```diff
@@ -134,11 +156,17 @@ public int size() {
...
     if (nestedProjections[pos] != null) {
-      return javaClass.cast(nestedProjections[pos].wrap(struct.get(positionMap[pos], StructLike.class)));
+      return javaClass.cast(nestedProjections[pos].wrap(struct.get(structPos, StructLike.class)));
     }
 
-    return struct.get(positionMap[pos], javaClass);
+    if (structPos != -1) {
+      return struct.get(structPos, javaClass);
+    } else {
+      return null;
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/util/StructProjection.java:168]: Since this relies on a `null` for `nestedProjections[pos]`, I think it should set `nestedProjections[pos]` to `null` in the constructor where `positionMap[pos]` is set to -1.

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java [hunk 23]
```diff
@@ -38,6 +38,12 @@ private MetadataColumns() {
       Integer.MAX_VALUE - 2, "_pos", Types.LongType.get(), "Ordinal position of a row in the source data file");
   public static final NestedField IS_DELETED = NestedField.required(
       Integer.MAX_VALUE - 3, "_deleted", Types.BooleanType.get(), "Whether the row has been deleted");
+  public static final NestedField SPEC_ID = NestedField.required(
+      Integer.MAX_VALUE - 4, "_spec_id", Types.IntegerType.get(), "Spec ID to which a row belongs to");
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/MetadataColumns.java:42]: Nit: no need for "to" at the end of the doc because it already uses "to which".

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/MetadataColumns.java:42]: I think I would also rephrase this because the row doesn't belong to a spec. How about "Spec ID used to track the data file containing a row"?

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java [hunk 24]
```diff
@@ -38,6 +38,12 @@ private MetadataColumns() {
...
+  // the partition column type is not static and depends on all specs in the table
+  public static final int PARTITION_COLUMN_ID = Integer.MAX_VALUE - 5;
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/MetadataColumns.java:44]: I think that new reserved columns need to be added to the spec. We should also make sure the spec notes the ranges that are reserved. I'm not sure if we did that or just added specific IDs.

File: core/src/main/java/org/apache/iceberg/util/PartitionUtil.java [hunk 25]
```diff
@@ -51,6 +56,20 @@ private PartitionUtil() {
...
 
+    // add _spec_id
+    idToConstant.put(
+        MetadataColumns.SPEC_ID.fieldId(),
+        convertConstant.apply(Types.IntegerType.get(), task.file().specId()));
+
+    // add _partition
+    if (partitionType != null && partitionType.fields().size() > 0) {
+      StructLike coercedPartition = coercePartition(partitionType, spec, partitionData);
+      idToConstant.put(MetadataColumns.PARTITION_COLUMN_ID, convertConstant.apply(partitionType, coercedPartition));
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/util/PartitionUtil.java:67]: This needs to call `convertConstant` on each partition value. The assumption in all of the implementations is that it is called on primitive values, not on structs.

File: core/src/main/java/org/apache/iceberg/util/PartitionUtil.java [hunk 26]
```diff
@@ -51,6 +56,20 @@ private PartitionUtil() {
...
+    } else if (partitionType != null) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/util/PartitionUtil.java:68]: It seems strange to check `partitionType != null` twice rather than nesting.

File: spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java [hunk 27]
```diff
@@ -132,6 +142,15 @@ protected InputFile getInputFile(String location) {
     return inputFiles.get(location);
   }
 
+  protected Map<Integer, ?> constantsMap(FileScanTask task, Schema readSchema) {
+    if (readSchema.findField(MetadataColumns.PARTITION_COLUMN_ID) != null) {
+      StructType partitionType = Partitioning.partitionType(table);
+      return PartitionUtil.constantsMap(task, partitionType, BaseDataReader::convertConstant);
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:148]: I'd prefer just to call the method rather than trying to optimize by not adding the partition entry.

File: spark3/src/test/java/org/apache/iceberg/spark/source/TestSparkCatalog.java [hunk 28]
```diff
@@ -31,7 +30,14 @@
...
+    if (parts.length == 2) {
+      TestTables.TestTable table = TestTables.load(parts[0]);
+      String[] metadataColumns = parts[1].split(",");
+      return new SparkTestTable(table, metadataColumns, false);
```
Comment (rdblue) [inline spark3/src/test/java/org/apache/iceberg/spark/source/TestSparkCatalog.java:37]: This is just a different way to pass metadata columns in? Is there a better way to test it?

File: spark3/src/test/java/org/apache/iceberg/spark/source/TestSparkMetadataColumns.java [hunk 29]
```diff
@@ -0,0 +1,192 @@
...
+        .commit();
+
+    List<Object[]> expected = ImmutableList.of(
+        row(0, row(null, null)),
+        row(1, row("b1", null)),
+        row(2, row("b1", 2)),
+        row(3, row(null, 2))
+    );
+    assertEquals("Rows must match", expected,
+        sql("SELECT _spec_id, _partition FROM `%s$_spec_id,_partition` ORDER BY _spec_id", TABLE_NAME));
```
Comment (rdblue) [inline spark3/src/test/java/org/apache/iceberg/spark/source/TestSparkMetadataColumns.java:157]: This should probably have a `TODO` comment to fix it when we can rely on Spark 3.2.

File: spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java [hunk 30]
```diff
@@ -155,6 +174,24 @@ protected static Object convertConstant(Type type, Object value) {
         return ByteBuffers.toByteArray((ByteBuffer) value);
       case BINARY:
         return ByteBuffers.toByteArray((ByteBuffer) value);
+      case STRUCT:
+        StructType structType = (StructType) type;
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:178]: Could this use an `InternalRowWrapper` instead?

File: core/src/main/java/org/apache/iceberg/util/PartitionUtil.java (hunk 25 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/util/PartitionUtil.java:67]: Okay, I see what's happening. This PR updates the conversion for Spark and because most callers only use `constantsMap(task, func)` there's no way to get a partition passed through. I'm a little uneasy about this, but since we need to convert to a specific class for the reader I don't see a good way around it.

File: spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java (hunk 30 above)
Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:178]: Nevermind, that goes the other direction.

**Review by rdblue — APPROVED** [review]
I noted a few minor things, but overall this is good. Please fix some of those things and then feel free to merge.

File: spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java (hunk 27 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:148]: I'd prefer calling this all the time too. However, this would mean we can no longer query tables with unknown transforms. We need to know all transforms to build a common partition type. That's why I don't set the partition column if it is not requested.

Any thoughts on this, @rdblue?

Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:148]: Can we leave unknown transforms out of the partition type instead? We can just ignore them if they're unknown?

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 3 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/Partitioning.java]: Fixed.

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java (hunk 11 above)
Comment (aokolnychyi) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java]: Resolving, no longer applies.

File: core/src/main/java/org/apache/iceberg/util/PartitionUtil.java (hunk 13 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/util/PartitionUtil.java]: I switched to using projections.

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 14 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/Partitioning.java]: Resolving this as this has been discussed.

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java (hunk 15 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MetadataColumns.java:46]: Added.

File: mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java (hunk 16 above)
Comment (aokolnychyi) [reply mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java]: No longer applies, resolving.

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java (hunk 17 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MetadataColumns.java]: Fixed.

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java (hunk 18 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MetadataColumns.java]: Fixed.

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 20 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/Partitioning.java]: Switched to projections, so no longer applies.

File: api/src/main/java/org/apache/iceberg/util/StructProjection.java (hunk 22 above)
Comment (aokolnychyi) [reply api/src/main/java/org/apache/iceberg/util/StructProjection.java:168]: I think it is actually always null as long as the field was not found. I'll add an explicit call, though.

File: spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java (hunk 27 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:148]: I'd probably consider persisting the partition type in the metadata instead. It might be confusing to silently ignore a partition column.

File: api/src/main/java/org/apache/iceberg/util/StructProjection.java (hunk 22 above)
Comment (aokolnychyi) [reply api/src/main/java/org/apache/iceberg/util/StructProjection.java:167]: Added.

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java (hunk 24 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MetadataColumns.java:44]: I'll do that in a follow-up as we are missing `_deleted` too.

File: spark3/src/test/java/org/apache/iceberg/spark/source/TestSparkMetadataColumns.java (hunk 29 above)
Comment (aokolnychyi) [reply spark3/src/test/java/org/apache/iceberg/spark/source/TestSparkMetadataColumns.java:159]: Added TODOs before all tests and to `SparkTestTable`.

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java (hunk 12 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MetadataColumns.java]: Fixed.

File: api/src/main/java/org/apache/iceberg/util/StructProjection.java (hunk 21 above)
Comment (aokolnychyi) [reply api/src/main/java/org/apache/iceberg/util/StructProjection.java]: Fixed.

File: core/src/main/java/org/apache/iceberg/MetadataColumns.java (hunk 23 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MetadataColumns.java]: Changed.

File: core/src/main/java/org/apache/iceberg/util/PartitionUtil.java (hunk 26 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/util/PartitionUtil.java]: Fixed.

File: core/src/main/java/org/apache/iceberg/util/PartitionUtil.java (hunk 25 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/util/PartitionUtil.java]: Yeah, it is also related to unknown transforms as well.

File: spark3/src/test/java/org/apache/iceberg/spark/source/TestSparkCatalog.java (hunk 28 above)
Comment (aokolnychyi) [reply spark3/src/test/java/org/apache/iceberg/spark/source/TestSparkCatalog.java:37]: I wanted to simply overload `newScanBuilder` in `SparkTestTable` but it did not work. We can get rid of this once we migrate to Spark 3.2 so I don't worry too much.

File: spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java (hunk 27 above)
Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:148]: I think it would be fine for unknown partitions right now. It would unblock this without much risk.

**Comment by aokolnychyi** [comment]
Thanks for reviewing, @rdblue @jackye1995 @RussellSpitzer @kbendick @stevenzwu!
